rfd/0048-tsh-sso-test-command.md

```diff
@@ -0,0 +1,170 @@
+URLs).
+
+This change of flow will allow for:
+
+- prompt closure of failed console logins - no need for timeouts or manual interruption of login
+- avenue for reporting of rich information about the cause of failure in console (in addition to Teleport debug logs)
+- keeping the existing error messages in the browser.
+
+### Implementation details
+
+There are several conceptual pieces to the implementation:
+
+1. Extending auth requests with embedded connector details. The `"/webapi/{oidc,saml,github}/login/console"` endpoints
+   will have to accept the additional parameter which subsequently will be stored with the auth request. There are
+   numerous types involved here which will need to be updated, along with conversions between them. For example, among
+   others, these types will need to be extended:
+
+- [`services.SAMLAuthRequest`](https://github.com/gravitational/teleport/blob/8c4bf751b211e82b555653a9aee6c6c5bf39411f/lib/services/identity.go#L421)
+- [`services.OIDCAuthRequest`](https://github.com/gravitational/teleport/blob/8c4bf751b211e82b555653a9aee6c6c5bf39411f/lib/services/identity.go#L352)
+- [`services.GithubAuthRequest`](https://github.com/gravitational/teleport/blob/8c4bf751b211e82b555653a9aee6c6c5bf39411f/lib/services/identity.go#L286)
+
+2. Making the backend aware of the testing flow. Right now there is no concept of "dry run" login flow, this needs to be
+   changed as appropriate. For example for SAML this will mean
+   updating [Server.ValidateSAMLResponse()](https://github.com/gravitational/teleport/blob/9b8b9d6d0c115d43d31d53c47db3050e27edbc4a/lib/auth/saml.go#L317)
+   , [Server.validateSAMLResponse()](https://github.com/gravitational/teleport/blob/9b8b9d6d0c115d43d31d53c47db3050e27edbc4a/lib/auth/saml.go#L361)
+   and others to skip creation of web sessions and signing the client public key.
+
+3. Extending the callback with debugging information and calling it on failed SSO login. This means updating the calling
+   logic (for SAML this
+   is [Handler.samlACS(()](https://github.com/gravitational/teleport/blob/940c83c16133fd9fc506f780e71e7b94edabf9d6/lib/web/saml.go#L89)
+   function) and receiving logic (
+   i.e. [Redirector.callback()](https://github.com/gravitational/teleport/blob/4db05acbef43847c4d899d54c3b1de2301234cc2/lib/client/redirect.go#L200))
+   .
+
+The implementation of this RFD for different kinds of connectors should be largely independent. As such, the first
+iteration will implement this functionality for SAML, while the lessons learned will help shape the implementations for
+OIDC and Github.
+
+### Enterprise version of `tsh`
+
+SAML and OIDC are Enterprise-only features. `tsh` needs to load the resources for these auth connectors and that is only
+possible using Enterprise version of the libraries. Similar to how `tctl` offers two tiers, `tsh` would follow the same
+pattern.
```

Comment:
That was the original idea :-) but you mentioned some Teleport Cloud based workflow that could be done benefit from adding this to `tsh` instead? But yeah, it can live with `tctl` just fine.

Reply:
Moved the command to `tctl`.